hs_core/views/utils.py

```diff
@@ -93,6 +106,158 @@ def upload_from_irods(username, password, host, port, zone, irods_fnames, res_fi
+    if curr_path == prefix_path or not curr_path.startswith(prefix_path):
+        filelist = add_resource_files(res_id, fileobj)
+    else:
+        curr_path = curr_path[len(prefix_path) + 1:]
+        filelist = add_resource_files(res_id, fileobj, folder=curr_path)
+
+    if filelist:
+        return filelist[0]
+    else:
+        return None
+
+
+def add_reference_url_to_resource(user, res_id, ref_url, ref_name, curr_path):
+    """
+    Add a reference URL to a composite resource if URL is valid, otherwise, return error message
+    :param user: requesting user
+    :param res_id: resource uuid
+    :param ref_url: reference url to be added to the resource
+    :param ref_name: file name of the referenced url in internet shortcut format
+    :param curr_path: the folder path in the resource to add the referenced url to
+    :return: 200 status code and 'success' message if it succeeds, otherwise, return error status
+    code and error message
+    """
+    ref_name = ref_name.lower()
+    if not ref_name.endswith('.url'):
+        ref_name += '.url'
+
+    is_valid, err_msg = validate_url(ref_url)
+    if not is_valid:
+        return status.HTTP_400_BAD_REQUEST, err_msg
+
+    f = add_url_file_to_resource(res_id, ref_url, ref_name, curr_path)
+
+    if not f:
+        return status.HTTP_500_INTERNAL_SERVER_ERROR, 'New url file failed to be added to the ' \
+                                                      'resource'
+
+    # make sure the new file has logical file set and is single file aggregation
+    try:
+        # set 'SingleFile' logical file type to this .url file
+        res = hydroshare.utils.get_resource_by_shortkey(res_id)
+        set_logical_file_type(res, user, f.id, 'SingleFile', extra_data={'url': ref_url})
+        hydroshare.utils.resource_modified(res, user, overwrite_bag=False)
+    except Exception as ex:
+        return status.HTTP_500_INTERNAL_SERVER_ERROR, ex.message
+
+    return status.HTTP_200_OK, 'success'
+
+
+def edit_reference_url_in_resource(user, res, new_ref_url, curr_path, url_filename):
+    """
+    edit the referenced url in an url file
+    :param user: requesting user
+    :param res: resource object that includes the url file
+    :param new_ref_url: new url to replace the old url
+    :param curr_path: the folder path of the url file in the resource
+    :param url_filename: the url file name in the resource
+    :return: 200 status code and 'success' message if it succeeds, otherwise, return error status
+    code and error message
+    """
+    ref_name = url_filename.lower()
+    if not ref_name.endswith('.url'):
+        return status.HTTP_400_BAD_REQUEST, 'url_filename in the request must have .url extension'
+    is_valid, err_msg = validate_url(new_ref_url)
+    if not is_valid:
+        return status.HTTP_400_BAD_REQUEST, err_msg
+
+    # temp path to hold updated url file to be written to iRODS
+    temp_path = os.path.join(getattr(settings, 'IRODS_ROOT', '/tmp'), uuid4().hex)
```

Comment:
I didn't know about this.  I think I'll borrow it for the overwrite.  One thought though, should IStorage have a temp directory method so irods logic doesn't leak into the main django app?

Reply:
maybe just add getTmpPath to IStorage that could read the settings IRODS_ROOT and create the path with /tmp and a random hex.  So the main django ap is unaware of this IRODS specific setting.  Once the path is returned, the main django app can create files in there and then delete the folder when finished?

Reply:
@sblack-usu OK. Will do. As we agreed, will call it ```getUniqueTmpPath()```